packages/protocol/contracts/L1/libs/LibVerifying.sol

```diff
@@ -196,41 +198,36 @@ library LibVerifying {
         uint32 _gasLimit = blk.gasLimit + LibL2Consts.ANCHOR_GAS_COST;
         assert(fc.gasUsed <= _gasLimit);
 
-        IProverPool proverPool =
-            IProverPool(resolver.resolve("prover_pool", false));
+        blk.verifiedForkChoiceId = fcId;
 
-        if (blk.assignedProver == address(0)) {
+        if (blk.prover == address(0)) {
             --state.slotB.numOpenBlocks;
-        } else if (!blk.proverReleased) {
-            proverPool.releaseProver(blk.assignedProver);
         }
 
-        // Reward the prover (including the oracle prover)
-        uint64 proofReward =
-            (config.blockFeeBaseGas + fc.gasUsed) * blk.rewardPerGas;
-
-        if (fc.prover == address(1)) {
-            // system prover is rewarded with `proofReward`.
-        } else if (blk.assignedProver == address(0)) {
-            // open prover is rewarded with more tokens
-            proofReward = proofReward * config.rewardOpenMultipler / 100;
-        } else if (blk.assignedProver != fc.prover) {
-            // proving out side of the proof window, by a prover other
-            // than the assigned prover
-            proofReward = proofReward * config.rewardOpenMultipler / 100;
-            proverPool.slashProver(blk.blockId, blk.assignedProver, proofReward);
-        } else if (fc.provenAt <= blk.proposedAt + blk.proofWindow) {
-            // proving inside the window, by the assigned prover
+        bool inProofWindow = blk.prover == address(0)
+            || fc.provenAt <= blk.proposedAt + blk.proofWindow;
+
+        // Calculate the block fee
+        uint32 feePerGas = inProofWindow
+            ? blk.feePerGas
+            : state.slotC.avgFeePerGas * config.rewardOpenMultipler / 100;
+
+        uint64 blockFee = LibUtils.calcBlockFee(config, fc.gasUsed, feePerGas);
+
+        // Update protocol level stats
+        if (inProofWindow && fc.prover != address(1)) {
             uint64 proofDelay;
             unchecked {
                 proofDelay = fc.provenAt - blk.proposedAt;
+            }
 
-                if (config.rewardMaxDelayPenalty > 0) {
-                    // Give the reward a penalty up to a small percentage.
-                    // This will encourage prover to submit proof ASAP.
-                    proofReward -= proofReward * proofDelay
-                        * config.rewardMaxDelayPenalty / 10_000 / blk.proofWindow;
-                }
+            if (config.rewardMaxDelayPenalty > 0) {
+                // Give the reward a penalty up to a small percentage.
+                // This will encourage prover to submit proof ASAP.
+                blockFee -= uint64(
+                    uint256(blockFee) * proofDelay
+                        * config.rewardMaxDelayPenalty / 10_000 / blk.proofWindow
+                );
```

Comment:
Similar to above I guess, not sure if we need to incentivize faster provers like this, because this is a tax on provers that want to aggregate as many blocks as possible. If the network guarantees that proofs only to be submitted within the proof window, some blocks being proven faster doesn't really help the network and this extra value being leaked here will end up with users.

Reply:
will remove as well